8281561: Disable http DIGEST mechanism with MD5 and SHA-1 by default

Michael-Mc-Mahon opened the pull request:
Hi,

Could I get the following change reviewed please, which is to disable the MD5 message digest algorithm by default in the HTTP Digest authentication mechanism? The algorithm can be opted into by setting a new system property "http.auth.digest.reEnabledAlgs" to include the value MD5. The change also updates the Digest authentication implementation to use some of the more secure features defined in RFC7616, such as username hashing and additional digest algorithms like SHA256 and SHA512-256.

- Michael

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed
- [x] Change requires a CSR request to be approved

### Issues
 * [JDK-8281561](https://bugs.openjdk.java.net/browse/JDK-8281561): Disable http DIGEST mechanism with MD5 and SHA-1 by default
 * [JDK-8282649](https://bugs.openjdk.java.net/browse/JDK-8282649): Disable http DIGEST mechanism with MD5 by default (**CSR**)


### Reviewers
 * [Weijun Wang](https://openjdk.java.net/census#weijun) (@wangweij - **Reviewer**) ⚠️ Review applies to c55fdd94c63fc7514842a82b3d65d351ccba56c9
 * [Daniel Fuchs](https://openjdk.java.net/census#dfuchs) (@dfuch - **Reviewer**) ⚠️ Review applies to b2095e02657e253f702917c33fefca334d247ba8


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7688/head:pull/7688` \
`$ git checkout pull/7688`

Update a local copy of the PR: \
`$ git checkout pull/7688` \
`$ git pull https://git.openjdk.java.net/jdk pull/7688/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7688`

View PR using the GUI difftool: \
`$ git pr show -t 7688`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7688.diff">https://git.openjdk.java.net/jdk/pull/7688.diff</a>

</details>

Michael-Mc-Mahon commented:
/csr needed

Michael-Mc-Mahon commented:
I have still to add the doc update describing the system property, which will come shortly. I may suggest a change of name to the system property to better reflect its exact meaning/purpose

dfuch commented:
Should we instead have a property to disable algorithms, whose default value would contain "MD5" by default?

Michael-Mc-Mahon commented:
> Should we instead have a property to disable algorithms, whose default value would contain "MD5" by default?

I considered that and implemented it that way at the start, but what you would end up with then is users running their code with something like: -DdisabledAlgNames="" 

I find that style leads to a much less explicit "opting in" than by making the user explicitly identify the deprecated algorithm by name.

dfuch commented:
> I considered that and implemented it that way at the start, but what you would end up with then is users running their code with something like: -DdisabledAlgNames=""
> 
> I find that style leads to a much less explicit "opting in" than by making the user explicitly identify the deprecated algorithm by name.

Right - but it would also allow user to opt-in to disable more algorithms by listing them in the property

Michael-Mc-Mahon commented:
> > I considered that and implemented it that way at the start, but what you would end up with then is users running their code with something like: -DdisabledAlgNames=""
> > I find that style leads to a much less explicit "opting in" than by making the user explicitly identify the deprecated algorithm by name.
> 
> Right - but it would also allow users to opt-in to disable more algorithms by listing them in the property

In practical terms, the only other likely candidate there is SHA-1. If that weren't the case, I'd disagree with your point.

So, maybe, we could have a 2nd net property with the default disabled algorithms and in net.properties we identify MD5 only for now. Users could add to that list if they want or even specify it on the command line. I think it's potentially confusing, but maybe there is a case for adding to the disabled list.

Michael-Mc-Mahon commented:
> > > I considered that and implemented it that way at the start, but what you would end up with then is users running their code with something like: -DdisabledAlgNames=""
> > > I find that style leads to a much less explicit "opting in" than by making the user explicitly identify the deprecated algorithm by name.
> > 
> > 
> > Right - but it would also allow users to opt-in to disable more algorithms by listing them in the property
> 
> In practical terms, the only other likely candidate there is SHA-1. If that weren't the case, I'd disagree with your point.
> 
> So, maybe, we could have a 2nd net property with the default disabled algorithms and in net.properties we identify MD5 only for now. Users could add to that list if they want or even specify it on the command line. I think it's potentially confusing, but maybe there is a case for adding to the disabled list. I need to think about a way to do this without subvertng the point about making the user explicitly opt in.



> > > I considered that and implemented it that way at the start, but what you would end up with then is users running their code with something like: -DdisabledAlgNames=""
> > > I find that style leads to a much less explicit "opting in" than by making the user explicitly identify the deprecated algorithm by name.
> > 
> > 
> > Right - but it would also allow users to opt-in to disable more algorithms by listing them in the property
> 
> In practical terms, the only other likely candidate there is SHA-1. If that weren't the case, I'd disagree with your point.
> 
> So, maybe, we could have a 2nd net property with the default disabled algorithms and in net.properties we identify MD5 only for now. Users could add to that list if they want or even specify it on the command line. I think it's potentially confusing, but maybe there is a case for adding to the disabled list. I need to think about a way to do this without subvertng the point about making the user explicitly opt in.

Thinking about it again, I wonder if we should just deprecate SHA-1 at the same time. I think there will be less compatibility impact than with MD5, and it's basically broken as well. I don't see a reason to opt out of other algorithms at this time.

dfuch commented:
> > So, maybe, we could have a 2nd net property with the default disabled algorithms and in net.properties we identify MD5 only for now. Users could add to that list if they want or even specify it on the command line. I think it's potentially confusing, but maybe there is a case for adding to the disabled list. I need to think about a way to do this without subvertng the point about making the user explicitly opt in.
> 
> Thinking about it again, I wonder if we should just deprecate SHA-1 at the same time. I think there will be less compatibility impact than with MD5, and it's basically broken as well. I don't see a reason to opt out of other algorithms at this time.

I see - maybe we should have a security property identifying the list of algorithm that are disabled, and then a system property to reenable them?

dfuch commented on src/java.base/share/classes/sun/net/www/protocol/http/DigestAuthentication.java:
```diff
@@ -60,19 +63,50 @@
+    // used for proxy connections, or plain text http server connections
+
+    // This will probably be expanded to include SHA-1 eventually
+    private static final Set<String> defDisabledAlgs =
+        Set.of("MD5");
```
What I'm suggesting is that the content of this set could be seeded with the value of a Security Property, defined in java.security - rather than have the default value hardcoded here.

Michael-Mc-Mahon replied:
Yes, I think that should work

jaikiran commented on src/java.base/share/classes/java/net/doc-files/net-properties.html:
```diff
@@ -224,6 +224,12 @@ <H2>Misc HTTP URL stream protocol handler properties</H2>
 	      property is defined, then its value will be used as the domain
 	      name.</P>
 	</OL>
+	<LI><P><B>{@systemProperty http.auth.digest.reEnabledAlgs}</B> (default: &lt;none&gt;)<BR>
```
Hello Michael, from what I understand of `doc-files` directory (which is where this html file resides) the `javadoc` tool considers it an unprocessed files location. So would adding the `systemProperty` javadoc taglet here be necessary? I think this won't end up being listed in the system properties index generated by the javadoc tool and I think this line here will just get rendered literally.

dfuch replied:
This actually seems to work. If you build the docs for JDK mainline, you can search for instance, for `http.keepAlive.time.proxy` in the javadoc search box and it will lead you `net-properties.html`.

jaikiran replied:
You are right indeed. I just checked the generated system properties index and it does list these properties and even links back to this document. In fact, I even found a mail which clearly states that this processing of `@systemProperties` is supported for `doc-files` https://mail.openjdk.java.net/pipermail/core-libs-dev/2018-November/056653.html

jaikiran commented on src/java.base/share/classes/sun/net/www/protocol/http/DigestAuthentication.java:
```diff
@@ -427,6 +508,24 @@ public void checkResponse (String header, String method, URL url)
+            if (logger.isLoggable(PlatformLogger.Level.INFO)) {
+                logger.info(msg + " This constraint may be relaxed by setting " +
+                     "the \"http.auth.digest.enabledDigestAlgs\" system property.");
+            }
```
I suspect this is a typo and perhaps should have been `http.auth.digest.reEnabledAlgs`?

Michael-Mc-Mahon replied:
Thanks. I'm going to change the name once more and will update it then.

jaikiran commented on src/java.base/share/classes/sun/net/www/protocol/http/DigestAuthentication.java:
```diff
@@ -378,9 +431,23 @@ private String getHeaderValueImpl (String uri, String method) {
+            validateAlgorithm(algorithm);
+        } catch (IOException e) {
+            // should not happen since the algorithm has already been
+            // validated
```
Hello Michael, was this comment meant for something else? The comments feels a bit odd since it says "has already been validated" which isn't the case since the `validateAlgorithm` itself has failed here.

Michael-Mc-Mahon replied:
Yes, copy and paste error. Thanks

jaikiran commented on src/java.base/share/classes/sun/net/www/protocol/http/DigestAuthentication.java:
```diff
@@ -471,31 +576,49 @@ public void checkResponse (String header, String method, String uri)
+    {
+        boolean truncate256 = false;
+
+        if (algorithm.equals("SHA-512-256")) {
```
It appears that the incoming param `algorithm` appears can be of any case. In some other places like the `validateAlgorithm`, this `algorithm` value has been first converted to an uppercase value and then used for additional checks. Should a similar upper case conversion be done here before this equality check? Perhaps, we should convert this to upper case once and then pass it around to these `validateAlgorithm` and `computeUserhash` methods?

Michael-Mc-Mahon replied:
Right. I'll check all that for the next round.

jaikiran commented on src/java.base/share/classes/sun/net/www/protocol/http/DigestAuthentication.java:
```diff
@@ -367,6 +419,7 @@ private String getHeaderValueImpl (String uri, String method) {
         String nonce = params.getNonce ();
         String algorithm = params.getAlgorithm ();
+        boolean userhash = params.getUserhash ();
         params.incrementNC ();
```
I am not familiar with the code and the request/response flow involved in this code, so what I mention may not be relevant. However, do you think the algorithm validation that we added in this PR should happen before this `incrementNC()` is called and the `ncString`  construction is done? I see that this incremented `ncCount` then gets used in the `checkResponse` part. In the case where the algorithm validation fails and we return `null` from this method (which effectively means not setting the authorization header), do you think the subsequent `checkResponse` would run into issues due to the `incrementNC()` being done here?

Michael-Mc-Mahon replied:
I think there probably wouldn't be a problem as the value would always be increasing and the server is only checking for duplicates, or replays rather than gaps. But, nevertheless, it seems like better practice to do the check before incrementing the counter.

wangweij commented on src/java.base/share/classes/sun/net/www/protocol/http/DigestAuthentication.java:
```diff
@@ -530,7 +653,8 @@ private String computeDigest(
-    private String encode(String src, char[] passwd, MessageDigest md) {
+    // truncate256 means only use the last 256 bits of the digest (32 bytes)
+    private String encode(String src, char[] passwd, MessageDigest md, boolean truncate256) {
         md.update(src.getBytes(ISO_8859_1.INSTANCE));
```
Maybe we can support the "charset" parameter as well. The only allowed value is "UTF-8".

Michael-Mc-Mahon replied:
I'll look into supporting that.

wangweij reviewed (commented):
Several comments added. Also, I am reading rfc7616 and it will be really nice if we can include the 2 examples at https://datatracker.ietf.org/doc/html/rfc7616#section-3.9 as a known answer test. Of course this means you need a way to provide a hardcoded `cnonce` and it's probably not easy.

jaikiran commented on src/java.base/share/classes/sun/net/www/protocol/http/DigestAuthentication.java:
```diff
@@ -427,6 +508,24 @@ public void checkResponse (String header, String method, URL url)
+    private void validateAlgorithm(String algorithm) throws IOException {
+        if (getAuthType() == AuthCacheValue.Type.Server &&
+                getProtocolScheme().equals("https")) {
+            // HTTPS server authentication can use any algorithm
```
Hello Michael,
Should it be noted somewhere in the text of the linked CSR that the property values play no role if the request URL's scheme is HTTPS (and proxy isn't involved)?

As a related question - should the implementation even bother about this HTTP vs HTTPS protocol check? As far as I understand the reason why we are proposing to disable support for MD5 is because of its cryptographic weakness/vulnerabilities. So irrespective of whether or not the  computed MD5 (or any other disabled algorithm) digest value gets transferred over HTTP or HTTPS, the value would still be cryptographically vulnerable isn't it?

Michael-Mc-Mahon replied:
The distinction may be worth mentioning in the CSR and the docs. The distinction is useful because like Basic authentication which is completely insecure with plaintext HTTP, it is secure and a potentially useful web authentication scheme controlled by the browser when run over HTTPS (except when authenticating with HTTPS to a proxy).

Michael-Mc-Mahon commented on src/java.base/share/classes/java/net/doc-files/net-properties.html:
```diff
@@ -224,6 +224,12 @@ <H2>Misc HTTP URL stream protocol handler properties</H2>
+        authentication due to their proven security limitations. Currently, this list only
+        includes {@code MD5} but other algorithms may be added in future. If it is still
+        required to use one of these algorithms, then they can be re-enabled by setting
+        this property to a comma separated list of the algorithm names.</P>
```
Right, probably better to allow whitespace, which seems to be commonly used in the existing security properties

Michael-Mc-Mahon commented on src/java.base/share/classes/sun/net/www/protocol/http/DigestAuthentication.java:
```diff
@@ -60,19 +63,50 @@
+
+    // A net property which overrides the disabled set above.
+    private static final String enabledAlgPropName = "http.auth.digest." +
+        "reEnabledAlgs";
```
I'll try and see if it fits the normal line width

Michael-Mc-Mahon commented on src/java.base/share/classes/sun/net/www/protocol/http/DigestAuthentication.java:
```diff
@@ -540,9 +664,14 @@ private String encode(String src, char[] passwd, MessageDigest md) {
+        int start = 0;
+        if (truncate256) {
+            assert digest.length >= 32;
+            start = digest.length - 32;
```
Okay, I'll double check that. I haven't found any server implementations of this feature to test with yet,

wangweij replied:
2nd test of https://datatracker.ietf.org/doc/html/rfc7616#section-3.9 is on this algorithm, but it requires UTF-8 charset support and a way to provide a predefined cnonce. If it's not worth modifying our implementation to create a regression test, I think at least we can temporarily hack our own JDK and try on it. And I think it's most likely true that this algorithm is using a different initialization vector as Bernd pointed out.

wangweij replied:
As https://www.rfc-editor.org/errata_search.php?rfc=7616&rec_status=0 shows, that 2nd test in rfc7616 was wrong in the initial version as it used a truncated version of SHA-512. The real SHA-512/256 algorithm used be used.
```
$ jshell
jshell> import java.security.MessageDigest

jshell> HexFormat.of().formatHex(MessageDigest.getInstance("SHA-512").digest("J\u00e4s\u00f8n Doe:api@example.org".getBytes("UTF-8")))
$2 ==> "488869477bf257147b804c45308cd62ac4e25eb717b12b298c79e62dcea254ec5211a6631b181289b4dd8c14890f38f04bff8a388106dabb900c6984ba592b6a"

jshell> HexFormat.of().formatHex(MessageDigest.getInstance("SHA-512/256").digest("J\u00e4s\u00f8n Doe:api@example.org".getBytes("UTF-8")))
$3 ==> "793263caabb707a56211940d90411ea4a575adeccb7e360aeb624ed06ece9b0b"
```

Michael-Mc-Mahon replied:
We could put a property in there to seed the random number generator, but I'd prefer not to do that. I will check with a modified JDK that it works with this data though.

seanjmullan commented on src/java.base/share/classes/sun/net/www/protocol/http/DigestAuthentication.java:
```diff
@@ -427,6 +508,24 @@ public void checkResponse (String header, String method, URL url)
+    private void validateAlgorithm(String algorithm) throws IOException {
+        if (getAuthType() == AuthCacheValue.Type.Server &&
+                getProtocolScheme().equals("https")) {
+            // HTTPS server authentication can use any algorithm
```
A more security conscious user may want to disable MD5 digest authentication, even when used over HTTPS, even though the risks are far less than when used over HTTP. Is there a way to do that?

Michael-Mc-Mahon replied:
There isn't a way to do that currently. I'd prefer not to further complicate the configuration to allow it. It would probably be a lot simpler to just disable MD5 across the board for proxy and server instead? We're including SHA-1 in that list now as well by the way.

dfuch commented on src/java.base/share/classes/sun/net/www/protocol/http/DigestAuthentication.java:
```diff
@@ -60,19 +64,39 @@
+            new PrivilegedAction<>() {
+                public String run() {
+                    return Security.getProperty(secPropName)
+                                   .replaceAll("\\s", "")
```
`Security.getProperty` may return `null` so replacement should only be made after checking that it is non null.

dfuch replied:
Maybe `String.trim()` should be called on each element after splitting instead: you want to remove spaces before and after commas, not necessarily spaces within a name. "MD 5, SHA-256" probably shouldn't be parsed as "MD5,SHA-256".

Michael-Mc-Mahon replied:
Okay.

dfuch commented on src/java.base/share/classes/sun/net/www/protocol/http/DigestAuthentication.java:
```diff
@@ -60,19 +64,39 @@
+        Set<String> disabledAlgs = new HashSet<>();
+        @SuppressWarnings("removal")
+        String secprops = AccessController.doPrivileged(
+            new PrivilegedAction<>() {
```
could use a lambda instead of an anonymous class?

Michael-Mc-Mahon replied:
Had tried it first and compiler didn't know whether lambda is a PrivilegedAction or a PrivilegedExceptionAction, but it seems a cast works. Will change it.

dfuch replied:
or declare a local variable that holds the lambda - that would avoid the cast.

dfuch commented on src/java.base/share/classes/sun/net/www/protocol/http/DigestAuthentication.java:
```diff
@@ -60,19 +64,39 @@
+        for (String algorithm : algorithms) {
+            disabledAlgs.add(algorithm);
+        }
+        return Collections.unmodifiableSet(disabledAlgs);
```
Could use `Set.copyOf()` ? Or is there anything that might call contains(null)? Also should strings be converted to upper case like below?

dfuch commented on src/java.base/share/classes/sun/net/www/protocol/http/DigestAuthentication.java:
```diff
@@ -84,23 +108,23 @@
                     String s = NetProperties.get(enabledAlgPropName);
-                    return s == null ? "" : s;
+                    return s == null
+                        ? "" : s.replaceAll("\\s", "").toUpperCase();
```
Should probably use Local.ROOT to convert to upper case.
It seems to me that the code that takes a String as argument, check for null and returns an empty set, remove spaces, convert it to upper case, splits the string at commas, and create an immutable set from that, could be moved to an auxillary function and called for parsing both the Security property and the System property - since their syntax is identical.

Michael-Mc-Mahon replied:
good idea

dfuch commented on src/java.base/share/classes/sun/net/www/protocol/http/DigestAuthentication.java:
```diff
@@ -361,6 +394,7 @@ public boolean setHeaders(HttpURLConnection conn, HeaderParser p, String raw) {
         params.setOpaque (p.findValue("opaque"));
         params.setQop (p.findValue("qop"));
         params.setUserhash (Boolean.valueOf(p.findValue("userhash")));
+        params.setCharset(p.findValue("charset", "ISO_8859_1").toUpperCase());
```
Should probably use Locale.ROOT when converting to upper case

dfuch commented on src/java.base/share/classes/sun/net/www/protocol/http/DigestAuthentication.java:
```diff
@@ -389,6 +423,7 @@ public boolean setHeaders(HttpURLConnection conn, HeaderParser p, String raw) {
         if (algorithm == null || algorithm.isEmpty()) {
             algorithm = "MD5";  // The default, accoriding to rfc2069
         }
+        algorithm = algorithm.toUpperCase();
```
same here

wangweij commented on src/java.base/share/classes/sun/net/www/protocol/http/DigestAuthentication.java:
```diff
@@ -389,6 +423,7 @@ public boolean setHeaders(HttpURLConnection conn, HeaderParser p, String raw) {
         if (algorithm == null || algorithm.isEmpty()) {
             algorithm = "MD5";  // The default, accoriding to rfc2069
         }
+        algorithm = algorithm.toUpperCase();
```
Please use `toUpperCase(Locale.ROOT)` or `toUpperCase(Locale.ENGLISH)`.

Michael-Mc-Mahon replied:
> Should we have a test case for "SHA-512-256" too?

If there's hardcoded example in the RFC, I will include it

Michael-Mc-Mahon commented on src/java.base/share/classes/sun/net/www/protocol/http/DigestAuthentication.java:
```diff
@@ -60,19 +64,39 @@
-    private static final String enabledAlgPropName = "http.auth.digest." +
-        "reEnabledAlgs";
+    private static final String enabledAlgPropName =
+        "http.auth.digest.enabledAlgorithms";
```
A net property can be a system property. But, it can also be specified in the net.properties file. We're using different names for the security and net property as the security property specifies algortithms to be disabled and the net property ones to be (re)enabled.

Michael-Mc-Mahon commented on src/java.base/share/classes/sun/net/www/protocol/http/DigestAuthentication.java:
```diff
@@ -198,6 +224,17 @@ synchronized boolean getUserhash() {
+            return "UTF-8".equals(charset)
+                ? StandardCharsets.UTF_8
+                : StandardCharsets.ISO_8859_1;
+        }
```
You mean reject the whole response as a protocol error? I guess we probably should do that.

wangweij commented on src/java.base/share/classes/sun/net/www/protocol/http/DigestAuthentication.java:
```diff
@@ -530,8 +685,11 @@ private String computeDigest(
+            src = Normalizer.normalize(src, Normalizer.Form.NFC);
+        }
+        md.update(src.getBytes(charset));
         if (passwd != null) {
```
[RFC7616 ](https://datatracker.ietf.org/doc/html/rfc7616#section-4) says the password should also support unicode. It's OK we use the old code for compatibility reason.

Michael-Mc-Mahon commented on src/java.base/share/conf/security/java.security:
```diff
@@ -701,6 +701,16 @@ jdk.security.legacyAlgorithms=SHA1, \
+# property "http.auth.digest.enabledAlgorithms" to a comma
+# separated list of algorithms to be allowed.
+#
+jdk.httpdigest.defaultDisabledAlgorithms = MD5, MD-5, SHA1, SHA-1
```
The aliases are a PITA. It's a shame there isn't support in the public API to canonicalize the names. But, what I will do is canonicalize the incoming strings from the HTTP server and then we only have to list canonical names in the properties.

Michael-Mc-Mahon commented on src/java.base/share/classes/java/net/doc-files/net-properties.html:
```diff
@@ -224,6 +224,14 @@ <H2>Misc HTTP URL stream protocol handler properties</H2>
+        usable for HTTPS server authentication. The default list of disabled algorithms is specified
+        in the {@code java.security} properties file and currently comprises {@code MD5} and
+        {@code SHA-1}. If it is still required to use one of these algorithms, then they can be
+        re-enabled by setting this property to a comma separated list of the algorithm names.</P>
```
Okay, I'm suggesting "http.auth.digest.reEnabledAlgorithms" now. 
Hopefully we can stick with that.

Michael-Mc-Mahon commented on src/java.base/share/classes/sun/net/www/protocol/http/DigestAuthentication.java:
```diff
@@ -57,12 +64,48 @@
+
+    private static final Set<String> defDisabledAlgs = getDefaultAlgs();
+
+    private static Set<String> getDefaultAlgs() {
```
That code is reworked so the method no longer exists

Michael-Mc-Mahon commented on src/java.base/share/classes/sun/net/www/protocol/http/DigestAuthentication.java:
```diff
@@ -540,7 +698,6 @@ private String encode(String src, char[] passwd, MessageDigest md) {
         }
         byte[] digest = md.digest();
-
         StringBuilder res = new StringBuilder(digest.length * 2);
```
The fix will probably be back ported, so I'd prefer not to use HexFormat.

Michael-Mc-Mahon commented:
I still plan to add a couple more tests for UTF-8 encoding and if possible for username hashing

wangweij commented on src/java.base/share/classes/sun/net/www/protocol/http/DigestAuthentication.java:
```diff
@@ -337,6 +427,12 @@ public boolean setHeaders(HttpURLConnection conn, HeaderParser p, String raw) {
         if (algorithm == null || algorithm.isEmpty()) {
             algorithm = "MD5";  // The default, accoriding to rfc2069
         }
+        var oid = KnownOIDs.findMatch(algorithm.toUpperCase(Locale.ROOT));
```
No need to call `toUpperCase`. `findMatch` already called that. On the other hand, we don't support the name `SHA-512-256`, and if you translate it to stdName here there is no need to do it again in `computeUserhash`.

wangweij reviewed (commented):
Please remove `src/java.base/share/classes/java/util/.Random.java.swp`.

wangweij reviewed (approved):
LGTM now. It will be even nicer if the known answer tests in RFC 7616 can be included.

Michael-Mc-Mahon commented on src/java.base/share/classes/sun/net/www/protocol/http/DigestAuthentication.java:
```diff
@@ -64,41 +71,35 @@
+        propPrefix + "reEnabledAlgorithms";
 
-    private static final Set<String> disabledAlgs = new HashSet<>();
+    private static final Set<String> disabledAlgorithms = new HashSet<>();
```
The Set is private to the class and is not modified after the static initializer completes.  It's not clear to me how using Set.copyOf provides strong MT-safe guarantees than this.

dfuch replied:
Better safe than sorry. An alternative could be to use ConcurrentHashMap.newKeySet(); But since it's supposed to be both immutable and MT-safe then Set.copyOf() would probably be a lighter and better choice.

Michael-Mc-Mahon commented:
webrevs 06 & 07 have the latest changes, if anyone is just looking at the incremental changes

dfuch commented on src/java.base/share/classes/sun/net/www/protocol/http/DigestAuthentication.java:
```diff
@@ -487,15 +521,12 @@ private String getHeaderValueImpl (String uri, String method) {
-        if (session) {
-            algorithm = algorithm.substring(0, algorithm.length() - 5);
-        }
+        boolean session = algorithm.endsWith ("-sess");
```
should that be `digest.endsWith("-sess");` ?

Michael-Mc-Mahon replied:
No. The new method `setAlgorithmNames` ensures that the suffix is in lower case in the parameters object.

Michael-Mc-Mahon replied:
No, the digest field refers to the actual message digest algorithm (as known to the security libraries). The algorithm field holds the algorithm name as it is defined in  RFC71616.

dfuch replied:
I am confused here - because you converted `algorithm` to upper case, so it should never end with `-sess`?

Michael-Mc-Mahon replied:
Look at line 478: The `algorithm` field is reset here to be the upper case of the digest name plus the -sess suffix in lower case.

dfuch replied:
Doh! Somehow I'd read that as `digestName = digestName + "-sess"`.
All clear now.

Michael-Mc-Mahon commented:
/integrate
